op-challenger/fault/position_test.go

```diff
@@ -1 +1,122 @@
 package fault
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestMSBIndex(t *testing.T) {
+	tests := []struct {
+		input    uint64
+		expected int
+	}{
+		{0, 0},
+		{1 << 0, 0},
+		{1 << 1, 1},
+		{1 << 2, 2},
+		{1 << 3, 3},
+		{16, 4},
+		{255, 7},
+		{1024, 10},
+		{18446744073709551615, 63},
+	}
+
+	for _, test := range tests {
+		result := MSBIndex(test.input)
+		if result != test.expected {
+			t.Errorf("MSBIndex(%d) expected %d, but got %d", test.input, test.expected, result)
+		}
+	}
+
+}
+
+func TestGINConvessions(t *testing.T) {
+	tests := []struct {
+		index    uint64
+		position Position
+	}{
+		{
+			index:    1,
+			position: Position{Depth: 0, IndexAtDepth: 0},
+		},
+		{
+			index:    0b10,
+			position: Position{Depth: 1, IndexAtDepth: 0},
+		},
+		{
+			index:    0b100,
+			position: Position{Depth: 2, IndexAtDepth: 0},
+		},
+		{
+			index:    0b110,
+			position: Position{Depth: 2, IndexAtDepth: 2},
+		},
+		{
+			index:    0b1010,
+			position: Position{Depth: 3, IndexAtDepth: 2},
+		},
+		{
+			index:    0b1100,
+			position: Position{Depth: 3, IndexAtDepth: 4},
+		},
+	}
+	for _, test := range tests {
+		from := FromGIN(test.index)
+		require.Equal(t, test.position, from)
+		to := test.position.ToGIN()
+		require.Equal(t, test.index, to)
+	}
+}
+
+func TestTraceIndex(t *testing.T) {
+	// Note: for whatever reason there appears to be an extra bit in the system & there are two valid options for every trace index.
+	// I think it is because we always go left at the start, though it is the lowest bit that we a re free to change without issue...
+	// Maybe it is fine to change that low bit because we always go left or do a loop that ends up going left so the lowest bit
+	// is always zero
+	tests := []struct {
+		pos Position
+		idx int
+	}{
+		{
+			pos: Position{Depth: 0, IndexAtDepth: 0},
+			idx: 8,
+		},
+		{
+			pos: Position{Depth: 1, IndexAtDepth: 0},
+			idx: 4,
+		},
+		{
+			pos: Position{Depth: 2, IndexAtDepth: 0}, // 0 or 1?
+			idx: 2,
+		},
+		{
+			pos: Position{Depth: 2, IndexAtDepth: 1}, // 0 or 1?
+			idx: 2,
+		},
+		{
+			pos: Position{Depth: 2, IndexAtDepth: 2}, // 2 or 3?
+			idx: 6,
+		},
```

Comment:
`IndexAtDepth` has to monotonically increase for each node at a specific depth. It doesn't really matter if it starts from 0 or 1 but there has to be a power of 2 number of nodes at each depth except 0 and 1 which are the weird bit before the actual binary tree gets started. So I think we have an extra node at depth 2 which should only have 2. Should be `Depth: 2, IndexAtDepth: 0 -> idx: 2` and `Depth: 2, IndexAtDepth: 1 -> idx: 6`. 

I was going to suggest we steal the tests from `LibPosition` but it only has fuzz tests so is just repeating the production logic in the tests and not really testing anything...

So I suspect we need to get a small, concrete example of a tree with the correct generalised indices, depth/indexAtDepth and trace index for each node. Then we can plug them into both this test and LibPosition tests to ensure the two implementations actually match.

Reply:
cc @clabby

Reply:
Yep - spot on with the first paragraph.

Can work on this after #6080. I have a PR open (#6081) to update some documentation, could add it there. Some of the fuzz tests (`testFuzz_defend_correctness` & `testFuzz_rightIndex_correctness`) are moreso differential tests, but others are just sanity checks.

Reply:
In my mind the tree looks like:
<img width="1194" alt="image" src="https://github.com/ethereum-optimism/optimism/assets/72675/79c69d65-f78a-44f2-a0f8-53d29876146b">
Where the `2,1` means depth 2, index 1. I'm not sure how you make generalised indices work in a non-binary tree like this but I'm guessing you pretend there's two children of the root and so wind up never using gIndex 3. So from depth, index to gIndex would be:

```
0,0 -> 1
1,0 -> 2
(1,1 should exist in a binary tree and would be gIndex 3, but can never exist in the game)
2,0 -> 4
2,1 -> 5
3,0 -> 6
3,1 -> 7
3,2 -> 8
3,3 -> 9
```

Reply:
I'm assuming in that tree that the trace index rounds up when it winds up at a non-integer value but rounding down would also work.

Reply:
Not quite - You are correct that gindex 3 cannot exist, but it's left child can (gindex = 6). If the initial move at gindex = 2 is defended, the counter claim will be made at gindex = 6.

If we want to be particular, the number of nodes in each depth in the game is `2^d - 1`. In the tree, it's `2^d`. Effectively, none of these nodes can exist in the game because they all commit to the same trace index that the root claim does:
<img width="1180" alt="Screenshot 2023-06-20 at 7 07 37 PM" src="https://github.com/ethereum-optimism/optimism/assets/8406232/d8a082e2-f39a-469e-89fe-c64fbe39f0db">

Depth to gIndex would be:
```
0,0 -> 1
1,0 -> 2
(1,1 should exist in a binary tree and would be gIndex 3, but can never exist in the game)
2,0 -> 4
2,1 -> 5
2,2 -> 6
2,3 -> 7
3,0 -> 8
3,1 -> 9
3,2 -> 10
3,3 -> 11
3,4 -> 12
3,5 -> 13
3,6 -> 14
3,7 -> 15
... and so on
```

Reply:
ahhh ok, there's a few concepts in there that I had wrong so this diagram is really useful. If we can add the trace indices onto that we'd have the data we need for the tests.

Reply:
That tree makes sense. It's been implemented with the high being 8 instead of 7, but I can change that. Maybe that also fixes the indexing.

I think what tripped me up with the generalized index <-> position conversion where the base cases. I think there might still be a bug in the position <-> trace index conversion.

Reply:
I would suggest when we're converting between position and trace index we add tests for all nodes in that binary tree, even if the game will never hit them. It makes things so much easier to understand.  Then we can have separate tests for how attack/defend should move in the tree which will demonstrate that some of those nodes can never actually be reached.

Reply:
hmm, I'm also pretty sure that 2,2 & 2,3 should be able to exist. Depth 0 has 0 bits for the index. Depth 1 has 1 bit for the index at depth (i.e. 2^1 = 0b10) & Depth 2 has 2 bits for the index (2^2 = 0b100)

Reply:
Just for my clarity / to align vocab: What do we mean when we say "trace index"?
1. The index at depth of the leaf node (i.e. for a tree of max depth = 4, `[0,2^4)`)
2. The gindex of the leaf node (i.e. for a tree of max depth = 4, `[16,32)`)

We plan on having the `rightIndex` function in `LibPosition` return the gindex of the leaf node (it currently returns the index at depth), but the index at depth can still be extracted from the gindex regardless.

Reply:
By "trace index" I mean the index in the trace that we should get the claim value from.

Reply:
Got it, so that'd be #1. In the tree I posted, the trace index by that definition would be `gIndex - depth << 1` (or more simply, the gindex with the most significant bit unset).

Reply:
Yes. I'd like to be very clear about what the trace index is for non-leaf nodes as well so that we know we're picking the right value for each claim.

Reply:
Right, so that's gIndex 6 has a trace index of 11 (max depth 4). Which is another difference to my previous understanding - I was thinking that the trace index was the middle of the descendant leaf nodes.  And that also explains why the attack/defend rules for movement are different to what I expect as well.

Reply:
So in Go style I see that tree as:
```

type testNodeInfo struct {
	GIndex     uint64
	Position   Position
	TraceIndex uint64
}

var treeNodesMaxDepth4 = []testNodeInfo{
	{GIndex: 1, Position: Position{Depth: 0, IndexAtDepth: 0}, TraceIndex: 15},

	{GIndex: 2, Position: Position{Depth: 1, IndexAtDepth: 0}, TraceIndex: 7},
	{GIndex: 3, Position: Position{Depth: 1, IndexAtDepth: 1}, TraceIndex: 15},

	{GIndex: 4, Position: Position{Depth: 2, IndexAtDepth: 0}, TraceIndex: 3},
	{GIndex: 5, Position: Position{Depth: 2, IndexAtDepth: 1}, TraceIndex: 7},
	{GIndex: 6, Position: Position{Depth: 2, IndexAtDepth: 2}, TraceIndex: 11},
	{GIndex: 7, Position: Position{Depth: 2, IndexAtDepth: 3}, TraceIndex: 31},

	{GIndex: 8, Position: Position{Depth: 3, IndexAtDepth: 0}, TraceIndex: 1},
	{GIndex: 9, Position: Position{Depth: 3, IndexAtDepth: 1}, TraceIndex: 3},
	{GIndex: 10, Position: Position{Depth: 3, IndexAtDepth: 2}, TraceIndex: 5},
	{GIndex: 11, Position: Position{Depth: 3, IndexAtDepth: 3}, TraceIndex: 7},
	{GIndex: 12, Position: Position{Depth: 3, IndexAtDepth: 4}, TraceIndex: 9},
	{GIndex: 13, Position: Position{Depth: 3, IndexAtDepth: 5}, TraceIndex: 11},
	{GIndex: 14, Position: Position{Depth: 3, IndexAtDepth: 6}, TraceIndex: 13},
	{GIndex: 15, Position: Position{Depth: 3, IndexAtDepth: 7}, TraceIndex: 15},

	{GIndex: 16, Position: Position{Depth: 4, IndexAtDepth: 0}, TraceIndex: 0},
	{GIndex: 17, Position: Position{Depth: 4, IndexAtDepth: 1}, TraceIndex: 1},
	{GIndex: 18, Position: Position{Depth: 4, IndexAtDepth: 2}, TraceIndex: 2},
	{GIndex: 19, Position: Position{Depth: 4, IndexAtDepth: 3}, TraceIndex: 3},
	{GIndex: 20, Position: Position{Depth: 4, IndexAtDepth: 4}, TraceIndex: 4},
	{GIndex: 21, Position: Position{Depth: 4, IndexAtDepth: 5}, TraceIndex: 5},
	{GIndex: 22, Position: Position{Depth: 4, IndexAtDepth: 6}, TraceIndex: 6},
	{GIndex: 23, Position: Position{Depth: 4, IndexAtDepth: 7}, TraceIndex: 7},
	{GIndex: 24, Position: Position{Depth: 4, IndexAtDepth: 8}, TraceIndex: 8},
	{GIndex: 25, Position: Position{Depth: 4, IndexAtDepth: 9}, TraceIndex: 9},
	{GIndex: 26, Position: Position{Depth: 4, IndexAtDepth: 10}, TraceIndex: 10},
	{GIndex: 27, Position: Position{Depth: 4, IndexAtDepth: 11}, TraceIndex: 11},
	{GIndex: 28, Position: Position{Depth: 4, IndexAtDepth: 12}, TraceIndex: 12},
	{GIndex: 29, Position: Position{Depth: 4, IndexAtDepth: 13}, TraceIndex: 13},
	{GIndex: 30, Position: Position{Depth: 4, IndexAtDepth: 14}, TraceIndex: 14},
	{GIndex: 31, Position: Position{Depth: 4, IndexAtDepth: 15}, TraceIndex: 15},
}
```

And I think we should be able to iterate through that data to test both GIndex<->Position conversion and Position<->TraceIndex conversions.

Reply:
Sol impl of the above test can be found here:
https://github.com/ethereum-optimism/optimism/blob/6153d03df17e005be3a4cb095c718974a8a6a6d7/packages/contracts-bedrock/contracts/test/LibPosition.t.sol#L157

Reply:
I've coded it to work with the following tree. Red means unreacheable. T is trace index. P is position. G is generalized index.
![download](https://github.com/ethereum-optimism/optimism/assets/11803341/b8fc69ef-9004-4f21-895a-fd4b6ccebbcc)

Reply:
I think we're getting ahead of ourselves thinking about how we can move around the tree.  For any node we should be able to convert between gIndex, position and trace index.  That's what we need to get working as a primitive construct - it doesn't matter if movement rules allow us to get to a node or not, the conversions should still work.

If we mix movement rules into the conversion logic or tests, then we tightly couple movement and conversion when they should be independent.

Reply:
Ok, one test case was wrong (GIndex: 7 should be trace index 15, not 31).
Otherwise tests are all passing.